0017-sampler-interface.md

```diff
@@ -0,0 +1,248 @@
+
+```python
+# Our first circuit has many Parameters, but elects to do all measurement at 
+# the end. Note that measure_all() chooses the register name "meas" for you.
+circuit1 = QuantumCircuit(9)
+...
+<use 2039 Parameters in this circuit>
+...
+circuit1.measure_all()
+
+# Specify 128 different parameter value sets for circuit1, in a 32x4 shape.
+parameter_values1 = np.random.uniform(size=(32, 4, 2039))
+
+# Our second circuit happens to have no parameters, but three output registers 
+# named alpha, beta, and gamma of various sizes
+alpha = ClassicalRegister(127, "alpha")
+beta = ClassicalRegister(10, "beta")
+gamma = ClassicalRegister(15, "gamma")
+circuit2 = QuantumCircuit(QuantumRegister(127), alpha, beta, gamma)
+...
+<insert gates, measure into the registers, apply dynamic circuit features, etc.>
+...
+
+# invoke a sampler
+sampler = Sampler(..., shots=1024)
+job = sampler.run([
+    (circuit1, parameter_values1),
+    circuit2
+])
+
+# wait for the result
+job.result()
+
+# get a bundle of results for every task
+>> PrimitiveResult(
+>>     SamplerTaskResult(
+>>         DataBundle(meas=BitArray(<num_samples=1024, num_bits=127, shape=(32, 4)>)), 
+>>         <metadata>
+>>     ), 
+>>     SamplerTaskResult(
+>>         DataBundle(
+>>             alpha=BitArray(<num_samples=1024, num_bits=127, shape=()>)
+>>             beta=BitArray(<num_samples=1024, num_bits=10, shape=()>)
+>>             gamma=BitArray(<num_samples=1024, num_bits=15, shape=()>)
+>>         ), 
+>>         <metadata>
+>>     ), 
+>>     metadata=<global_metadata>
+>> )
+```
+
+The above example shows various features of the design.
+Notice the output results are placed in a `DataBundle`, a namespace container described [here](https://github.com/Qiskit/RFCs/pull/53), with one entry for each output of the corresponding task's circuits.
+These circuits only have bit-array outcomes, as allowed by Qiskit, and therefore all results have type `BitArray`, described later.
+For now, consider the following usage examples.
+
+```python
+
+# get result from the second task above
+result = job.result()[1]
+
+# get the entire quasi-probability distribution
+result.get_quasiprobabilities()
+>> QuasiProbabilityDistribution({1351235663: 0.0009765625, 325: 0.0009765625, 45262346723465234523453: 0.0009765625, 35435345235234372457453: 0.0009765625, ...})
+
+# marginalize a couple of indexes in the beta output
+result.data.beta.get_counts(marginalize_idxs=[0, 1])
+>> Counts({"00": 189, "10": 11, "01": 801, "11": 123})
+
+# flip all of the bits in the gamma according to some other compatibly-shaped bit-array
+flips = BitArray(my_flips, num_bits=15)
+flipped_gamma = result.data.gamma ^ flips
+```
+
+If a circuit had a non bit-array outcome, it would use a different type.
+We suggest just using `numpy.ndarrays` with `dtype=numpy.float64` for all non-int types like `angle[20]` for simplicity, at least initially, but we can cross that bridge when we get there.
```

Comment:
Well, `angle[20]` is a fixed-point type, so I would actually want to store that with an integer dtype.

Reply:
Okay, I can change it to a float type. Best not to make any unneeded contentious comments anyways.

Reply:
[f9e316b](https://github.com/ihincks/qiskit-rfcs/pull/9/commits/f9e316b2818d2a808f72d8cf508e7190ce34e8f9)